xmtp/src/session.rs

```diff
@@ -59,19 +56,50 @@ impl SessionManager {
         into: &EncryptedMessageStore,
     ) -> Result<Vec<u8>, SessionError> {
         let res = self.session.decrypt(&message)?;
-        let session_bytes = self.session_bytes()?;
         // TODO: Stop mutating/storing the persisted session and just build on demand
-        self.persisted.vmac_session_data = session_bytes;
-        self.persisted.save(into)?;
+        self.save(into)?;
 
         Ok(res)
     }
 }
 
 impl Store<EncryptedMessageStore> for SessionManager {
     fn store(&self, into: &EncryptedMessageStore) -> Result<(), StorageError> {
-        self.persisted.store(into)?;
-        Ok(())
+        StoredSession::try_from(self)?.store(into)
+    }
+}
+
+impl Save<EncryptedMessageStore> for SessionManager {
+    fn save(&self, into: &EncryptedMessageStore) -> Result<(), StorageError> {
+        StoredSession::try_from(self)?.save(into)
+    }
+}
+
+impl TryFrom<&StoredSession> for SessionManager {
+    type Error = StorageError;
+    fn try_from(value: &StoredSession) -> Result<Self, StorageError> {
+        let pickle = serde_json::from_slice(&value.vmac_session_data)
+            .map_err(|_| StorageError::SerializationError)?;
+
+        Ok(Self::new(
+            OlmSession::from_pickle(pickle),
+            value.peer_installation_id.clone(),
+        ))
+    }
+}
+
+impl TryFrom<&SessionManager> for StoredSession {
+    type Error = StorageError;
+
+    fn try_from(value: &SessionManager) -> Result<Self, Self::Error> {
+        Ok(StoredSession::new(
+            value.session.session_id(),
+            value.peer_installation_id.clone(),
+            // TODO: Better error handling approach. StoreError and SessionError end up being dependent on eachother
```

Comment:
What if we had `StorageError` convert serde serialization errors using `thiserrror`, so that we don't need to map the error type (and lose the error message).

```
    #[error("serialization error")]
    Serialization(#[from] serde_json::Error),
```